### ISSUE TITLE
fn:apply with last two arguments (array, map) for the positional and keyword args in a func-call

### ISSUE THREAD

**Issue by dnovatchev** [qt4cg/qtspecs#146]
The existing [**_fn:apply() in XPath 3.1_**](https://www.w3.org/TR/xpath-functions-31/#func-apply) "Makes a dynamic call on a function with an argument list supplied in the form of an array."

However in XPath 4.0 **_[a function may have both (required) positional and variadic arguments, expressed as  keyword arguments if the function is map-variadic](https://qt4cg.org/specifications/xquery-40/xpath-40.html#id-function-calls)_**. Thus there are cases when it is impossible to provide the arguments of the function just within an array.

**For such cases, we need a new overload**:

`fn:apply($function as function(*), $array as array(*), $map as map(*)) as item()*`
  
  **The result** of the function is obtained by creating and invoking the same dynamic call that would be the result of  a  function-call to `$function` with positional arguments taken from the (ordered) members of the supplied array `$array` and keyword arguments taken from the (unordered) entries (KVPs - keyword-value pairs) of  `$map`. 
  
 **The effect of calling** `fn:apply($f, [$a, $b, $c, ...], map{"k1" : v1, "k2" : v2, ...})` is the same as the effect of the dynamic function call resulting from `$function($a, $b, $c, ...., $k1 = v1, $k2 = v2, ...)`. The function conversion rules are applied to the supplied arguments in the usual way.
 
**Dependencies** 

1. The new overload of `fn:apply()` is needed **_[if we have in XPath 4.0 function calls with both positional and keyword arguments](https://qt4cg.org/specifications/xquery-40/xpath-40.html#id-function-calls)_**. In other words, this proposal supposes (depends on) the existence of this new XPath feature

2. Some newly proposed features, such as **_[Function Decorators](https://github.com/qt4cg/qtspecs/issues/106)_**, depend on the existence of this new overload of `fn:apply()`

**Example**:

The expression:

```xq
let $data := <a b="3"/>
 return
    fn:apply(fn:serialize#2, [$data], map{"method":"xml", "omit-xml-declaration":true()})
```

returns:

`'<a b="3"/>'`

**Comment by michaelhkay** [comment]
This is dependent on issue #160, which in turn is dependent on changing the data model and type system for dynamic function items. The current state of play is that the parameters of a function (item) have names (though the names are nowhere exposed) but they cannot be defined as optional and they don't carry a default value.

**Comment by dnovatchev** [comment]
> This is dependent on issue #160, which in turn is dependent on changing the data model and type system for dynamic function items. The current state of play is that the parameters of a function (item) have names (though the names are nowhere exposed) but they cannot be defined as optional and they don't carry a default value.

Yes, at present `fn:apply()` only covers the 3.1 case. This issue is that we need not to forget to add the corresponding overload(s) for the new ways function calls will be implemented for variadic functions, thus supplying not just an array of fixed number of arguments (as in Xpath 3.1) but also an array that contains the variadic arguments or a map that contains the map-variadic arguments.

**Comment by michaelhkay** [comment]
I propose that we close this until such time as we have function items with call-by-name capability. Currently dynamic calls on function items can specify arguments by position only, and until that changes, no extensions to fn:apply() are appropriate.

**Comment by dnovatchev** [comment]
> I propose that we close this until such time as we have function items with call-by-name capability. Currently dynamic calls on function items can specify arguments by position only, and until that changes, no extensions to fn:apply() are appropriate.

If we have a static function with optional (keyword-value) parameters, it would be really unfortunate not to allow a partial application of this function also to have optional parameters.

This is just one case where the need for optional parameters (of the resulting dynamic function) are necessary.

Yes, it would be really bad when making a partial application of a static function with optional parameters, to lose its capability of having any optional parameters.

This would force the developer to set all values for the optional arguments even before producing the partial application, but in some cases calculating the wanted argument values may depend on the logic after producing the partial application.

Similar case with assignment of a static function to a variable or with closures.

**Comment by michaelhkay** [comment]
Yes, but we have a separate issue (or possibly 2) asking for keyword parameters on dynamic function calls, and we can't consider this issue until that one is resolved one way or the other. I don't think it makes sense to have open issues that only become an issue if the specification is changed.

**Comment by dnovatchev** [comment]
> Yes, but we have a separate issue (or possibly 2) asking for keyword parameters on dynamic function calls, and we can't consider this issue until that one is resolved one way or the other. I don't think it makes sense to have open issues that only become an issue if the specification is changed.

Then why not have a label: "postponed" ?

Anyway, we will need to arrive at definite decisions for having  keyword parameters on dynamic function calls.

For me the distinction between "static" and "dynamic" becomes artificial and not useful when we have a partial application of a static function (why should this be dynamic?) or just assignment of a static function to a variable (even a stronger same question!).

**Comment by ndw** [comment]
The CG agreed to close this issue without further action at [meeting 077](https://qt4cg.org/meeting/minutes/2024/05-14.html).
